src/extension.ts

```diff
@@ -34,31 +34,51 @@ export function activate(context: vscode.ExtensionContext) {
 
     context.subscriptions.push(disposable1);
 
-    const disposable2 = vscode.commands.registerCommand('ucl-rsdg.validateAtf', async () => {
-        const editor = vscode.window.activeTextEditor;
-        const fileName = basename(editor.document.uri.fsPath);
-        const fileContent = editor.document.getText();
-        let fileProject: string;
-        try {
-            fileProject = getProjectCode(fileContent);
-        } catch (err) {
-            vscode.window.showErrorMessage(`Could not validate: ${err}`);
-            return;
-        }
-        // The validate function is currently not mapped to the appropriate logging functions
-        const result = await validate(fileName,fileProject,fileContent);
-        handleResult(result, editor);
+    const disposable2 = vscode.commands.registerCommand('ucl-rsdg.validateAtf', () => {
+        workWithServer("validate", validate);
         });
 
     context.subscriptions.push(disposable2);
 
+    context.subscriptions.push(
+        vscode.commands.registerCommand('ucl-rsdg.lemmatiseAtf', () => {
+            workWithServer("lemmatise", lemmatise)
+        })
+    );
+
     context.subscriptions.push(
         vscode.commands.registerCommand('ucl-rsdg.arabicPreview', () => {
             PreviewPanel.createOrShow(context.extensionUri);
         })
     );
 }
 
+/**
+ * Base function for performing server-related tasks with the editor contents.
+ *  
+ * Will first get some basic information from the editor (filename, project
+ * code, text contents), then call another function to communicate with
+ * the server, and finally display the results of that communication.
+ *
+ * @param verb a textual description of the action, for reporting purposes
+ * @param callback the function to call after getting the editor information
+ */
+async function workWithServer(verb: string, callback: ServerFunction): Promise<void> {
+    const editor = vscode.window.activeTextEditor;
+    const fileName = basename(editor.document.uri.fsPath);
+    const fileContent = editor.document.getText();
+    let fileProject: string;
+    try {
+        fileProject = getProjectCode(fileContent);
+    } catch (err) {
+        vscode.window.showErrorMessage(`Could not ${verb}: ${err}`);
+        return;
+    }
+    // The validate function is currently not mapped to the appropriate logging functions
+    const result = await callback(fileName,fileProject,fileContent);
```

Comment:
Shouldn't there be here something with a timeout configuration on this? Or something that checks the server is not responsive due to internet connection, server being down, etc? Maybe promises handle  this on their own, but I would expect this `await` to have a try/catch wrapper.

Reply:
That happens in the client code, and is caught there. I'll double check that nothing propagates to this point though (No, promises don't handle that on their own)

Reply:
Actually, even though errors are caught, there is still an issue, because we don't return a proper result from `validate` or `lemmatise`. So that logic needs a little reworking! This may have been related to the semi-random CI failures too.